From f6ebf25e0f73b9dde602a13cff7c1912290e770a Mon Sep 17 00:00:00 2001
From: "kirk.wang" <kirk.wang@amlogic.com>
Date: Tue, 16 May 2023 16:53:26 +0800
Subject: [PATCH 3/8] gst: CB1 add event AML-SET-MAX-BYTE-SIZE [1/1]

PD#SWPL-118787

Problem:
ts video data is more than setted multiqueue buffer size
at the beginning which will stuck the pipeline when multiqueue is filled

Solution:
tsdemux seend a event to multiqueue
when multiqueue get this event, set the buffer size to default 10M.

Verify:
t7c

Change-Id: I768c2870b559268073b79902b599120469eb1edf
Signed-off-by: kirk.wang <kirk.wang@amlogic.com>
---
 ...mux-send-event-AML-SET-MAX-BYTE-SIZE.patch | 45 +++++++++++++++++
 ...e-handle-event-AML-SET-MAX-BYTE-SIZE.patch | 50 +++++++++++++++++++
 .../gstreamer1.0-plugins-bad_%.bbappend       |  1 +
 .../gstreamer/gstreamer1.0_%.bbappend         |  1 +
 4 files changed, 97 insertions(+)
 create mode 100755 recipes-multimedia/gstreamer/gst1-plugins-bad/0012-tsdemux-send-event-AML-SET-MAX-BYTE-SIZE.patch
 create mode 100755 recipes-multimedia/gstreamer/gst1/0005-multiqueue-handle-event-AML-SET-MAX-BYTE-SIZE.patch

diff --git a/recipes-multimedia/gstreamer/gst1-plugins-bad/0012-tsdemux-send-event-AML-SET-MAX-BYTE-SIZE.patch b/recipes-multimedia/gstreamer/gst1-plugins-bad/0012-tsdemux-send-event-AML-SET-MAX-BYTE-SIZE.patch
new file mode 100755
index 00000000..0df4a8c2
--- /dev/null
+++ b/recipes-multimedia/gstreamer/gst1-plugins-bad/0012-tsdemux-send-event-AML-SET-MAX-BYTE-SIZE.patch
@@ -0,0 +1,45 @@
+From e2033c9436ed06cefbc538b6c615cee42a3b03a8 Mon Sep 17 00:00:00 2001
+From: "sheng.liu" <sheng.liu@amlogic.com>
+Date: Tue, 13 Dec 2022 15:22:36 +0800
+Subject: [PATCH] gstreamer1.0-plugins-bad: CB1 send event to multiqueue to
+ large buffer size [1/1]
+
+PD#SWPL-104970
+
+Problem:
+ts video data is more than setted multiqueue buffer size
+at the beginning which will stuck the pipeline when multiqueue is filled
+
+Solution:
+tsdemux seend a event to multiqueue
+when multiqueue get this event, set the buffer size to default 10M.
+
+Verify:
+AH212
+---
+ gst/mpegtsdemux/tsdemux.c | 9 +++++++++
+ 1 file changed, 9 insertions(+)
+
+diff --git a/gst/mpegtsdemux/tsdemux.c b/gst/mpegtsdemux/tsdemux.c
+index 44f2105..d857961 100644
+--- a/gst/mpegtsdemux/tsdemux.c
++++ b/gst/mpegtsdemux/tsdemux.c
+@@ -1774,6 +1774,15 @@ done:
+     gst_pad_set_caps (pad, caps);
+     gst_pad_set_query_function (pad, gst_ts_demux_srcpad_query);
+     gst_pad_set_event_function (pad, gst_ts_demux_srcpad_event);
++
++    if (is_video) {
++      GstStructure *s;
++      GstEvent *event;      
++      s = gst_structure_new_empty ("AML-SET-MAX-BYTE-SIZE");
++      event = gst_event_new_custom (GST_EVENT_CUSTOM_DOWNSTREAM_STICKY, s);
++      GST_DEBUG("Send SET-MAX-BYTE-SIZE Event");
++      gst_pad_push_event (pad, event);
++    }
+   }
+ 
+   g_free (name);
+-- 
+2.25.1
+
diff --git a/recipes-multimedia/gstreamer/gst1/0005-multiqueue-handle-event-AML-SET-MAX-BYTE-SIZE.patch b/recipes-multimedia/gstreamer/gst1/0005-multiqueue-handle-event-AML-SET-MAX-BYTE-SIZE.patch
new file mode 100755
index 00000000..0db89112
--- /dev/null
+++ b/recipes-multimedia/gstreamer/gst1/0005-multiqueue-handle-event-AML-SET-MAX-BYTE-SIZE.patch
@@ -0,0 +1,50 @@
+From b3093323fdc4e436976c97bb3b1e50b78c3d1fc9 Mon Sep 17 00:00:00 2001
+From: "sheng.liu" <sheng.liu@amlogic.com>
+Date: Tue, 13 Dec 2022 15:41:44 +0800
+Subject: [PATCH] gstreamer1.0-plugins-bad: CB1 multiqueue handle event
+ AML-SET-MAX-BYTE-SIZE  [1/1]
+
+PD#SWPL-104970
+
+Problem:
+ts video data is more than setted multiqueue buffer size
+at the beginning which will stuck the pipeline when multiqueue is filled
+
+Solution:
+tsdemux seend a event to multiqueue
+when multiqueue get this event, set the buffer size to default 10M.
+
+Verify:
+AH212
+---
+ plugins/elements/gstmultiqueue.c | 12 ++++++++++++
+ 1 file changed, 12 insertions(+)
+
+diff --git a/plugins/elements/gstmultiqueue.c b/plugins/elements/gstmultiqueue.c
+index d4feb89..bcb55c5 100644
+--- a/plugins/elements/gstmultiqueue.c
++++ b/plugins/elements/gstmultiqueue.c
+@@ -1672,6 +1672,20 @@ gst_single_queue_push_one (GstMultiQueue * mq, GstSingleQueue * sq,
+         /* Applying the gap may have made the queue non-full again, unblock it if needed */
+         gst_data_queue_limits_changed (sq->queue);
+         break;
++      case GST_EVENT_CUSTOM_DOWNSTREAM_STICKY:
++        if(gst_event_has_name(event, "AML-SET-MAX-BYTE-SIZE")) {
++          GST_DEBUG_OBJECT (mq, "Handle event AML-SET-MAX-BYTE-SIZE");
++          GST_MULTI_QUEUE_MUTEX_LOCK (mq);
++          mq->max_size.bytes = DEFAULT_MAX_SIZE_BYTES;
++          SET_CHILD_PROPERTY (mq, bytes);
++          mq->max_size.time = DEFAULT_MAX_SIZE_TIME * 4;
++          SET_CHILD_PROPERTY (mq, time);
++          GST_MULTI_QUEUE_MUTEX_UNLOCK (mq);
++          gst_multi_queue_post_buffering (mq);
++          gst_event_unref(event);
++          result = GST_FLOW_OK;
++        }
++        break;
+       default:
+         break;
+     }
+-- 
+2.25.1
+
diff --git a/recipes-multimedia/gstreamer/gstreamer1.0-plugins-bad_%.bbappend b/recipes-multimedia/gstreamer/gstreamer1.0-plugins-bad_%.bbappend
index e760394e..0e40661e 100755
--- a/recipes-multimedia/gstreamer/gstreamer1.0-plugins-bad_%.bbappend
+++ b/recipes-multimedia/gstreamer/gstreamer1.0-plugins-bad_%.bbappend
@@ -8,6 +8,7 @@ SRC_URI_append = "file://0006-get-more-data-if-offset-is-larger-than-buffer-size
 SRC_URI_append = "file://0007-modify-bug-fragment-loss.patch "
 SRC_URI_append = "file://0008-add-new-registration-id-444f5649-for-tsdemux.patch "
 SRC_URI_append = "file://0009-add-gap-time-of-pcr-to-last-pcr.patch "
+SRC_URI_append = "file://0012-tsdemux-send-event-AML-SET-MAX-BYTE-SIZE.patch "
 
 DEPENDS += "gst-aml-drmbufferpool-plugins"
 
diff --git a/recipes-multimedia/gstreamer/gstreamer1.0_%.bbappend b/recipes-multimedia/gstreamer/gstreamer1.0_%.bbappend
index ec7492e3..4e42515d 100644
--- a/recipes-multimedia/gstreamer/gstreamer1.0_%.bbappend
+++ b/recipes-multimedia/gstreamer/gstreamer1.0_%.bbappend
@@ -3,4 +3,5 @@ SRC_URI_append = "file://0001-gstreamer-add-poll-api.patch "
 SRC_URI_append = "file://0002-modfiy-free-sequence-to-resolve-video-switch-problam.patch "
 SRC_URI_append = "file://0003-gstreamer-clcok-adjust-issue.patch "
 SRC_URI_append = "file://0004-do-not-post-buffering-in-multiqueue.patch "
+SRC_URI_append = "file://0005-multiqueue-handle-event-AML-SET-MAX-BYTE-SIZE.patch "
 SRC_URI_append = "file://0006-gstreamer1.0-CF1-add-get-outstandingbufnum-interface.patch "
-- 
2.25.1

